Preview Bubble fix crash 2

aosyatnik opened the pull request:
### Purpose

NB! Fix is not working for Dynamo Revit. Crash is fixed, but you won't see ". . ." during computation.

Link to YouTrack:
[MAGN-9568](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-9568) Crash in preview control

# Reason of crash:
In Dynamo there are several threads, let's call them UI and Core. When there is big calculation, UI thread tries to get some value from core, but this value has not been calculated yet.
Look next image.
Here we generate list 0..1000000. MirrorData is generated, but when we try to get value from svData (StackValue), it returns null.
![image](https://cloud.githubusercontent.com/assets/8158404/13455892/e0d12ca0-e068-11e5-9853-b8b9f9107920.png)
But if we wait for some time until calculation is done, we won't get any exception.
![image](https://cloud.githubusercontent.com/assets/8158404/13455965/390d3a62-e069-11e5-8526-edc8e22e523f.png)

After some investigation, it was found, that there are 3 async tasks: 
* `UpdateGraphAsyncTask` (engine calculates values)
* `QueryMirrorDataAsyncTask` (cached value of node is updated)
* `DelegateBasedAsyncTask` (update preview bubble)

When there is big calculation, these tasks are called in wrong order.

![image](https://cloud.githubusercontent.com/assets/8158404/13527255/818fa172-e216-11e5-87f4-9c27bbb36dc6.png)
![image](https://cloud.githubusercontent.com/assets/8158404/13527260/8a2f7ff0-e216-11e5-9299-875294b649e4.png)

You can easily see it in the next image.
![image](https://cloud.githubusercontent.com/assets/8158404/13527739/f55b38b6-e219-11e5-878e-869738bf3721.png)

# Fixes:

* Priority of `DelegateBasedAsyncTask` is reduced to `BelowNormal`.

* Created class-stub `MirrorDataInProgress`, instance of which is set during graph calculation.

* Created `busyTextBlock`, which is shown in preview bubble during big calculation.

### Declarations

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.

### Reviewers

@ke-yu 
@Randy-Ma 
@Benglin 
@pbidenko 

### FYIs

@Racel 
@jnealb 

One thing I didn't fix, it's next situation:
![image](https://cloud.githubusercontent.com/assets/8158404/13456682/07c3415a-e06d-11e5-83f0-a78e4e74bafc.png)

It's said: "Run completed." But preview bubble still shows ". . .". That's because CachedValue for node is generated, but it still requires to be translated into user-friendly representation. We should created separate task for this, if we want to improve performance.

Created separate task for performance improvements: http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-9621

ke-yu commented on src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs:
```diff
@@ -307,6 +323,22 @@ private void RefreshCondensedDisplay(Action refreshDisplay)
+            if (mirrorData is ProtoCore.MirrorDataInProgress
+                // When CBN is created, its cached value is set to MirrorDataInProgress.
+                // And it's not updated after update graph task is complited.
+                && !((nodeViewModel.NodeLogic is CodeBlockNodeModel) && (nodeViewModel.NodeLogic as CodeBlockNodeModel).Code == String.Empty))
```
I think the crash happens on all nodes (though other kinds of nodes are not so easy to reproduce), not only to code block node.

Randy-Ma replied:
If it is only limited to code block nodes, the conditions are not written correctly. The 1st condition looks good enough.

ke-yu commented on src/Engine/ProtoCore/Reflection/MirrorDataInProgress.cs:
```diff
@@ -0,0 +1,27 @@
+    /// WARNING: this class is used as temp value for CachedValue in NodeModel.
+    /// Don't try to get some value, in any case it will be something wrong.
+    /// </summary>
+    public class MirrorDataInProgress : MirrorData
```
Can we move this class to somewhere else? Engine is a separate component, it shouldn't have much knowledge about Dynamo

aosyatnik commented on src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs:
```diff
@@ -307,6 +324,22 @@ private void RefreshCondensedDisplay(Action refreshDisplay)
+            if (mirrorData is MirrorDataInProgress
+                // When CBN is created, its cached value is set to MirrorDataInProgress.
+                // And it's not updated after update graph task is complited.
+                && (nodeViewModel.NodeLogic as CodeBlockNodeModel == null || (nodeViewModel.NodeLogic as CodeBlockNodeModel).Code != String.Empty))
```
@ke-yu , @Randy-Ma this check is not useless. It's situation, when CBN is just created. If we won't have this check, we will get the next situation:
![image](https://cloud.githubusercontent.com/assets/8158404/13630026/a2bb3154-e5e5-11e5-8270-0b137501dcf7.png)

Benglin commented on src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs:
```diff
@@ -49,6 +52,19 @@ private struct Element
+            Opacity = 0.5,
+            Margin = new System.Windows.Thickness(12, 6, 0, 10)
+        };
+
```
Instead of `TextBlock`, can we not just use a `MirrorData` that contains a `String` that is `...`, and use the regular watch view to show it? That way we don't have to manage another `TextBlock` just for this.

aosyatnik replied:
We can use string `MirrorData`. But in future it would be harder to change it into animated text as it's said in this task: http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-9584

Benglin replied:
Hi @aosyatnik, I have had a chat with @kronz, it seems like that's not a good enough justification for us to complicate the code further. So when @Racel has a chance to talk that through with Zach, we'll probably just close that task out :)

aosyatnik commented:
Hi, @Benglin , as you've asked: I set `CachedValue` as string `MirrorData` and removed `TextBlock`.

Also I removed call on Scheduler from `RefreshCondensedDisplay`. There is no any big calculation in `RefreshCondensedDisplay`, it only takes `mirrorData` and translates it into string for condensed view.
E.g. for 0..1000000 it's just "List".

Now we don't have crash and frozen UI at least for condensed view.

I had to comment 
`//tree.treeView1.ItemContainerGenerator.StatusChanged += WatchContainer_StatusChanged;`
for now. Because it causes one more crash, which I'm going to fix next... But it's not critical functionality.
With this commented code I disable this functionality:
![image](https://cloud.githubusercontent.com/assets/8158404/13641399/edfdb7f0-e620-11e5-8fdc-6b808cef950a.png)

When you collapse list item - preview bubble stays the same height. 

@Benglin , @ke-yu let's get it fixed :)

ke-yu commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -485,6 +485,23 @@ internal MirrorData GetCachedValueFromEngine(EngineController engine)
+        internal void SetCachedValueInProgress(ProtoCore.Core core, ProtoCore.RuntimeCore runtimeCore)
+        {
+            CachedValue =
+                cachedValueInProgress ?? new MirrorData(core, runtimeCore, core.Heap.AllocateFixedString(Configurations.BusyString));
```
seems `cachedValueInProgress` is always null

ke-yu commented on src/DynamoCore/Scheduler/DelegateBasedAsyncTask.cs:
```diff
@@ -13,7 +13,7 @@ public class DelegateBasedAsyncTask : AsyncTask
         public override TaskPriority Priority
         {
-            get { return TaskPriority.Normal; }
+            get { return TaskPriority.BelowNormal; }
```
Not to downgrade the priority of `DelegateBasedAsyncTask`.
